homeassistant/components/fing/coordinator.py

```diff
@@ -0,0 +1,66 @@
+"""DataUpdateCoordinator for Fing integration."""
+
+from datetime import timedelta
+import logging
+from random import randrange
+from typing import Self
+
+from fing_agent_api import FingAgent
+from fing_agent_api.models import Device
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import AGENT_IP, AGENT_KEY, AGENT_PORT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FingDataFetcher:
+    """Keep data from the Fing Agent."""
+
+    def __init__(self, hass: HomeAssistant, ip: str, port: int, key: str) -> None:
+        """Initialize Fing entity data."""
+        self._hass = hass
+        self._fing = FingAgent(ip, port, key)
+        self._network_id = None
+        self._devices: dict[str, Device] = {}
+
+    def get_devices(self):
+        """Return all the devices."""
+        return self._devices
+
+    def get_network_id(self) -> str | None:
+        """Return the network id."""
+        return self._network_id
+
+    async def fetch_data(self) -> Self:
+        """Fecth data from Fing."""
+        response = await self._fing.get_devices()
+        self._network_id = response.network_id
+        self._devices = {device.mac: device for device in response.devices}
+        return self
```

Comment:
why do we have this class?

Reply:
Primarily to keep the code cleaner, the coordinator manages the cadence of the updates, while the fetcher handles the actual update.
If you prefer, I can move everything inside the coordinator class.

Reply:
I think we can indeed merge this into the data update coordinator, it doesn't add that much imo